lib/source.js

```diff
@@ -20,17 +20,26 @@ const url = require('url');
 const SOURCE_BASENAME = 'source';
 
 class Source {
-    constructor(paramsSource, directory="") {
+    constructor(paramsSource, directory="", basename) {
+        // optional params for watermarking
+        if (paramsSource.watermarkContent) {
+            this.isWatermark = true;
+            this.watermarkContent = paramsSource.watermarkContent;
+            this.widthPercent = paramsSource.widthPercent;
```

Comment:
As mentioned in [my previous review](https://jira.corp.adobe.com/browse/NUI-806), this is watermark operation specific logic and should be fully decoupled from a generic source download code here. Otherwise we keep updating this whenever we add new operations like watermark that need file content, which makes it hard to maintain. Only the post processing logic (currently in `image.js`) should know about fields like `watermarkContent`.

Also important is that such extra files are only downloaded on demand when it actually gets to post processing and needs them. Not just automatically up front.

Instead we need to have

1. an API to download sources/files, based around the existing `Source` but generic so it can be used for other cases (IIUC basically the local source filename needs to be customizable by the consumer of the API)
   - the input is the string or object as used by `params.source` and what `watermarkContent` would also use as "type" to describe what to need
   - (nothing about `widthPercent` in there, that isn't needed for the reusable download part)
2. existing `getSource()` in `storage.js` uses it (like today)
3. `image.js` uses this API to download the watermark only when it actually needs it
   - it passes in the object, e.g. `rendition.instructions.watermarkContent`, so that the API never has to know about the names of those fields and it can be reused for different ones in the future
   - we might have to think about reuse of the same watermark between different renditions; maybe use the same filename then (but we need to ensure same url = same filename, different url = different filename)